File: nest-ts-bun-backend/tests/error-handling.test.ts
import express from 'express';
import request from 'supertest';
import { describe, expect, it } from 'vitest';
import { errorHandler } from '../src/middleware/error-handler';
import { notFoundHandler } from '../src/middleware/not-found';

describe('error handling', () => {
  it('returns 404 for unknown routes', async () => {
    const app = express();
    app.use(notFoundHandler);
    app.use(errorHandler);

    const response = await request(app).get('/missing');

    expect(response.status).toBe(404);
    expect(response.body.status).toBe(404);
    expect(response.body.code).toBe('NOT_FOUND');
    expect(response.body.title).toBe('Not Found');
    expect(response.body.type).toContain('/not-found');
    expect(response.body.requestId).toBeDefined();
  });

  it('returns normalized 500 payload', async () => {
    const app = express();

    app.get('/boom', () => {
      throw new Error('boom');
    });

    app.use(errorHandler);

    const response = await request(app).get('/boom');

    expect(response.status).toBe(500);
    expect(response.body.status).toBe(500);
    expect(response.body.code).toBe('INTERNAL_SERVER_ERROR');
    expect(response.body.detail).toBe('Unexpected server error.');
    expect(response.body.title).toBe('Internal Server Error');
    expect(response.body.requestId).toBeDefined();
  });
});

File: django-python-backend/tests/error_handling_test.py
def test_not_found_problem_details(client):
    response = client.get('/missing')
    assert response.status_code == 404
    payload = response.json()
    assert payload['status'] == 404
    assert payload['code'] == 'NOT_FOUND'


def test_internal_error_normalized(client):
    response = client.get('/api/v1/ping?fail=true')
    assert response.status_code == 500
    payload = response.json()
    assert payload['status'] == 500
    assert payload['code'] == 'INTERNAL_SERVER_ERROR'
    assert payload['traceId']
